proposals/nnnn-actors.md

```diff
@@ -140,6 +140,265 @@ extension BankAccount {
+
+This is exemplified by the following piece of code, exercising the `decisionMaker` actor:
+
+```swift
+async let shouldBeGood = decisionMaker.thinkOfGoodIdea() // runs async
+async let shouldBeBad = decisionMaker.thinkOfBadIdea() // runs async
+
+await shouldBeGood // could be .goodIdea or .badIdea ☠️
+await shouldBeBad
+```
+
+> This issue is illustrated by using async lets, however also simply manifest by more than 1 actor calling out to the same decision maker; one invoking `thinkOfGoodIdea` and the other one `thinkOfBadIdea`, a reentrant actor is not protecting us from such race conditions, making the programming model unnecessarily hard to reason about.
+
+This snippet _may_ result (depending on timing of the resumptions) in the following execution:
+
+```swift
+opinion = .goodIdea         // <1>
+// suspend: await tell(...) // <2>
+opinion = .badIdea          // | <4> (!)
+// suspend: await tell(...) // | <5>
+// resume: await tell(...)  // <2>
+return opinion              // <3>
+// resume: await tell(...)  // <5>
+return opinion              // <6>
+```
+
+But it _may_ also result in the "naively expected" execution, i.e. without interleaving, making the problem even trickier, because as with normal race conditions in concurrent code -- the issue will only show up when exercised in more real usage patterns, rather than early on in unit testing.
+
+With this example we have showcased that reentrant actors, by design, do not prevent "high-level" race conditions. They only prevent "low-level" data-races, as in: concurrent access to some variable they protect etc.
+
+Please note that the example here a simple single variable state, yet already it can catch people off guard because how similar it looks to synchronous code. The problem is greatly exaggerated in real applications where actors are used to isolate and protect complex state transitions. Because actors effectively can only communicate through async calls to other parts of the system, a typical actor function will not have one or two suspension points but many of them - leading to exponential growth of the possible states an actor function may observe during it's execution.
+
+> Existing actor implementations err on the side of non-reentrant actors by default, allowing for reentrancy to be optionally opted-into.
+>
+> - Erlang/Elixir ([gen_server](https://medium.com/@eduardbme/erlang-gen-server-never-call-your-public-interface-functions-internally-c17c8f28a1ee)) - showcases a simple "loop/deadlock" scenario and how to detect and fix it,
+> - Orleans ([Grains](https://dotnet.github.io/orleans/docs/grains/reentrancy.html)) offer extensive declarative configuration around reentrancy, it is also the closest to the Swift actors model,
```

Comment:
https://dotnet.github.io/orleans/docs/grains/reentrancy.htm - This is a very good read on the same issue and their solutions there; it matches up with how our actors look and behave like pretty well, so I'd recommend giving it a look

Reply:
Orleans seems to allow recursive re-entrancy, at least.

Reply:
Yeah, I think it fits our model as well as basically by saying "within the same task" reentrance can happen which is like their "Reentrancy within a call chain".

Let me add this to our proposal as well